### DISCUSSION TITLE
BilateralFilter cuda implementation seems to fail on inputs with >16 channels?

### DISCUSSION THREAD

**Discussion by ajabri** [Project-MONAI/MONAI#2083] [Q&A]
It seems that the BilateralFilter cuda kernel silently fails (returns zeros) on inputs with more than 16 channels. Here is an example of the 2D case:

```
# 16 channels
BilateralFilter.apply(torch.randn(10, 16, 60, 60).cuda().contiguous()).std()
tensor(1.0003, device='cuda:0')

# 17 channels
BilateralFilter.apply(torch.randn(10, 17, 60, 60).cuda().contiguous()).std()
tensor(0., device='cuda:0')
```

Is this expected?

Thanks!
Allan

**Comment by ajabri** [comment]
I wrote an extension of the kernel to the joint bilateral filter case that I'd like to contribute, just wanted to address this issue first :)

**Reply by wyli** [reply]
@charliebudd could you please help? 
thanks!

**Reply by charliebudd** [reply]
We should be throwing an error for such inputs, my bad. The kernels are optimised for each channel configuration however in doing so we must compile for each configuration, the number of configurations is capped to keep compile time down. We have a JIT system in the works which will fix this issue by being able to dynamically compile the cases requested by the users code. However, in the mean time, I will make a PR to explain the limitation in an error. 

I'm happy and interested to disscus your use case and extention @ajabri

**Reply by ajabri** [reply]
I see, got it!

I am using your BilateralFilter implementation on intermediate activations in a standard ResNet. In my case, I actually wanted a joint bilateral filter (e.g. https://johanneskopf.de/publications/jbu/paper/FinalPaper_0185.pdf), where the features for computing similarity may be different from input being filtered. For example, I am filtering the intermediate activations (i.e. 128dim spatial features), using the input image as a guide (i.e. 3dim RGB features).

Is there an easy way for me to extend the BilateralFilter to greater than 16 channels, at the cost of more compile time?

**Reply by ajabri** [reply]
I saw that the PHL-based filter allows for similar behaviour, but since my guide features are low-dim, I wanted to avoid using an approximation.

**Reply by charliebudd** [reply]
In the case of a bilateral filter, the PHL approximation is pretty much indistinguishable in my experience. If you're working with kernel sizes (spatial sigmas) greater than 8 ish (may differ for your hardware) then I would suggest the phl implementation as it really is much faster for practically no loss in quality. (As it happens the Bilateral filter currently defaults to this implementation any way. There is a boolean flag passed to the apply function if you wish to use the exact implementation)

As your use case is a joint bilateral filter then I would again suggest using the PHL implementation as the API allows for that already, to construct a bilateral filter just prepare you're feature vectors by concatinating the RGB components and the XY pixel coordinates.

Thankfully increasing the cap (which also applys to the PHL) shouldn't be an issue, 16 is a pretty arbitrary choice but as we dont want to increase our compile time I would suggest doing this in a fork. Like I said, we plan on introducing a JIT system and will move this code accross once it's in which will remove this issue in the future.

**Reply by ajabri** [reply]
I see, I'll try out the PHL implementation then. I've been using my own fork of the BilateralFilter cuda kernel (which allows for another feature tensor as input).

To increase the cap past 32, do I need to define new macros? Is there an easier way?

 Thanks for your help!

**Reply by charliebudd** [reply]
The template function is explicitly instantiated for all configuration (capped at 16 data channels and 19 feature channels) here...

https://github.com/Project-MONAI/MONAI/blob/dde11000a7a10bcc9ccae76570cbb0e92ea23cf9/monai/csrc/filtering/permutohedral/permutohedral_cuda.cu#L534-L537

you could just add your specific instatiations below e.g. 

`template void PermutohedralCuda<float, 128, 5>(float* values, float* positions, int elementCount, bool accurate);`

Performance may not be too desirable as I didn't optimise for such usage, I'll try to take this into account when I come back to refactor this for the JIT system.

**Reply by ajabri** [reply]
Got it, thanks!

Does the PHL-backed implementation of the BilateralFilter support backpropagation? I noticed the PHLFilter backward seems to have been commented out.

**Reply by charliebudd** [reply]
Ahh yes I should of thought about this given your use case. The backpropergation needs some looking at, I actually intended to get this straightened out this week.  I'll make an issue to track. Do you need gradients for the feature vectors and sigmas or just for the input being filtered?

**Comment by GHJg7** [comment]
hmm

**Comment by wyli** [comment]
The original post is 2 years old, more recently there's an extended implementation of bilateral/joint bilateral filtering https://github.com/Project-MONAI/MONAI/blob/65cf5fe1234d0d01cc52c9ab48395a3efbad6f45/monai/networks/layers/filtering.py#L345
